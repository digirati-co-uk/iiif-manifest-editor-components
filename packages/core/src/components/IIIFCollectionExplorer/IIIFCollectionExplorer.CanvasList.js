import * as React from 'react';
import * as ReactDOM from 'react-dom';
import * as PropTypes from 'prop-types';
import { Droppable, Draggable } from 'react-beautiful-dnd';
import {
  withStyles,
  withTheme,
  Typography,
  CircularProgress,
  Divider,
} from '@material-ui/core';
import { LazyImage } from 'react-lazy-images';
import { List } from 'react-virtualized';
import ContainerDimensions from 'react-container-dimensions';

import LocaleString from '../LocaleString/LocaleString';
import { getCanvasThumbnail } from './IIIFCollectionExplorer.utils';
import IMG from './IIIFCollectionExplorer.IMG';

const tileSize = 100;

const styles = theme => {
  const spacer = theme.spacing.unit;
  return {
    droppable: {
      display: 'block',
      flex: 1,
      outline: 0,
    },
    draggable: {
      outline: 0,
    },
    list: {
      outline: 0,
    },
    listItem: {
      display: 'flex',
      flexDirection: 'row',
      alignItems: 'flex-start',
      flexWrap: 'wrap',
      padding: `${spacer}px ${spacer}px 0 ${spacer}px`,
    },
    imageHolder: {
      width: tileSize,
      height: tileSize,
    },
    divider: {
      width: '100%',
      marginTop: spacer,
    },
    rowContent: {
      flex: 1,
      padding: spacer,
    },
    thumbnail: {
      objectFit: 'contain',
      width: tileSize,
      height: tileSize,
    },
    placeholder: {
      display: 'flex',
      width: tileSize,
      height: tileSize,
      alignItems: 'center',
      justifyContent: 'center',
    },
  };
};

const PortalAwareThumbnail = ({
  provided,
  snapshot,
  thumbnail,
  useLazy,
  canvas,
  classes,
}) => {
  const usePortal = snapshot.isDragging;
  const child = (
    <div
      ref={provided.innerRef}
      {...provided.draggableProps}
      {...provided.dragHandleProps}
    >
      {useLazy ? (
        <LazyImage
          src={thumbnail}
          alt={canvas.id}
          placeholder={({ imageProps, ref }) => (
            <div ref={ref} className={classes.placeholder}>
              <CircularProgress />
            </div>
          )}
          actual={({ imageProps }) => <img {...imageProps} />}
        />
      ) : (
        <IMG
          src={thumbnail}
          alt={canvas.id}
          className={classes.thumbnail}
          size={tileSize}
        />
      )}
    </div>
  );

  if (snapshot.isDragging) {
    //TODO this is just temporary.
    window.draggedData = JSON.parse(JSON.stringify(canvas));
  }

  if (!usePortal) {
    return child;
  }
  const portal = document.querySelector('.drag-drop-portal');
  // if dragging - put the item in a portal
  return ReactDOM.createPortal(child, portal);
};

class CanvasList extends React.Component {
  rowRenderer = ({
    // Unique key within array of rows
    key,
    // Index of row within collection
    index,
    // The List is currently being scrolled
    isScrolling,
    // This row is visible within the List (eg it is not an overscanned row)
    isVisible,
    // Style object to be applied to row (to position it)
    style,
  }) => {
    const { classes, manifestId } = this.props;
    const canvas = this.props.items[index];
    // TODO: intelligent pre-fetcher to figure out is there sizes in info JSON,
    // if there's no sizes try request a size fit in
    // if that doesn't work fallback for the full image
    const [thumbnail, useLazy] = getCanvasThumbnail(canvas);

    return (
      <div key={key} className={classes.listItem} style={style}>
        <div className={classes.imageHolder}>
          <Draggable
            key={canvas.id}
            draggableId={`${manifestId}||||${canvas.id}`}
            index={index}
            className={classes.draggable}
          >
            {(provided, snapshot) => (
              <PortalAwareThumbnail
                provided={provided}
                snapshot={snapshot}
                thumbnail={thumbnail}
                useLazy={useLazy}
                canvas={canvas}
                classes={classes}
              />
            )}
          </Draggable>
        </div>
        <div className={classes.rowContent}>
          <Typography component="h5" variant="h5">
            <LocaleString>{canvas.label}</LocaleString>
          </Typography>
          <Typography variant="subtitle1" color="textSecondary">
            {canvas.type}
          </Typography>
        </div>
        <Divider className={classes.divider} />
      </div>
    );
  };

  render() {
    const { classes, items, theme, droppableId } = this.props;
    return (
      <Droppable droppableId={droppableId} isDropDisabled={true}>
        {providedDroppable => (
          <div
            ref={providedDroppable.innerRef}
            {...providedDroppable.droppableProps}
            className={classes.droppable}
          >
            <ContainerDimensions>
              {({ width, height }) => (
                <List
                  width={width}
                  height={height}
                  rowCount={items.length}
                  rowHeight={theme.spacing.unit * 2 + tileSize + 1}
                  rowRenderer={this.rowRenderer}
                  className={classes.list}
                />
              )}
            </ContainerDimensions>
            {providedDroppable.placeholder}
          </div>
        )}
      </Droppable>
    );
  }
}

CanvasList.defaultProps = {
  items: [],
  droppableId: 'iiifimagelist',
};

export default withStyles(styles)(withTheme()(CanvasList));
