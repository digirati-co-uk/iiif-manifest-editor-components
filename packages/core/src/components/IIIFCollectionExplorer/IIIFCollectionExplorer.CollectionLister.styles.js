export default theme => ({
  list: {
    display: 'flex',
    flexWrap: 'wrap',
    overflowX: 'hidden',
    overflowY: 'auto',
    position: 'relative',
  },
  listItem: {
    display: 'block',
    width: '100%',
    cursor: 'pointer',
    display: 'flex',
    flexDirection: 'row',
    flexWrap: 'wrap',
    alignItems: 'flex-start',
    justifyContent: 'flex-start',
  },
  info: {
    width: 'calc(100% - 140px)',
    padding: theme.spacing.unit,
  },
  infoLong: {
    width: 'calc(100% - 40px)',
    padding: theme.spacing.unit,
  },
  moreVert: {
    width: 40,
    minHeight: 80,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
  thumbnail: {
    width: 100,
  },
  placeholder: {
    display: 'flex',
    width: 100,
    height: 100,
    alignItems: 'center',
    justifyContent: 'center',
  },
  divider: {
    width: '100%',
  },
});
