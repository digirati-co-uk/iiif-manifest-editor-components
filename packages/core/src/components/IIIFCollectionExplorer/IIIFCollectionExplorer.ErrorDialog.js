import * as React from 'react';
import * as PropTypes from 'prop-types';
import {
  Dialog,
  DialogTitle,
  DialogContent,
  DialogContentText,
  DialogActions,
  Button,
  Slide,
} from '@material-ui/core';

const ErrorDialog = ({ error, handleCloseErrorDialog }) => !!error && (
    <Dialog
      open={!!error}
      TransitionComponent={props => <Slide direction="up" {...props} />}
      onClose={handleCloseErrorDialog}
      aria-labelledby="error"
    >
      <DialogTitle>Error</DialogTitle>
      <DialogContent>
        <DialogContentText>{error}</DialogContentText>
      </DialogContent>
      <DialogActions>
        <Button onClick={handleCloseErrorDialog} color="primary">
          Ok
        </Button>
      </DialogActions>
    </Dialog>
  );

ErrorDialog.propTypes = {
  error: PropTypes.string,
  handleCloseErrorDialog: PropTypes.func,
};

ErrorDialog.defaultProps = {
  handleCloseErrorDialog: () => {},
};

export default ErrorDialog;