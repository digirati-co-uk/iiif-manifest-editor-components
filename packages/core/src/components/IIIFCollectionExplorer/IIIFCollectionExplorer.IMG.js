import * as React from 'react';
import { BrokenImage } from '@material-ui/icons';

// Internal image representation
class IMG extends React.Component {
  state = {
    error: false,
  };

  onError = () => this.setState({ error: true });

  render() {
    const { size, ...props } = this.props;
    return this.state.error ? (
      <div
        style={{
          width: size,
          height: size,
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'center',
        }}
      >
        <BrokenImage color="error" />
      </div>
    ) : (
      <div className="loader">
        <img {...props} onError={this.onError} />
      </div>
    );
  }
}

IMG.defaultProps = {
  size: 100,
};

export default IMG;
