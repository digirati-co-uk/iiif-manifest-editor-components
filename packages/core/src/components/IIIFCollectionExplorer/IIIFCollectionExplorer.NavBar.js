import * as React from 'react';
import * as PropTypes from 'prop-types';
import {
  withStyles,
  TextField,
  InputAdornment,
  Select,
  MenuItem,
  IconButton,
} from '@material-ui/core';
import { History, ArrowForward } from '@material-ui/icons';
import DefaultTooltip from '../DefaultTooltip/DefaultTooltip';

const style = theme => ({
  head: {},
  headAddressLine: {
    width: '100%',
    display: 'flex',
    flexDirection: 'row',
    padding: theme.spacing.unit,
    overflow: 'hidden',
    alignItems: 'center',
    justifyContent: 'flex-start',
  },
  urlInputStyles: {
    width: '100%',
  },
  loadResourceIconSpacer: {
    width: 50,
    marginLeft: 8,
  },
  selectHideout: {
    width: '100%',
    height: 0,
    overflow: 'hidden',
  },
  historyIconHolder: {
    cursor: 'pointer',
  },
});

class NavBar extends React.Component {
  state = {
    historyOpen: false,
  };

  handleHistoryOpen = () => {
    this.setState({ historyOpen: true });
  };

  handleHistoryClose = () => {
    this.setState({ historyOpen: false });
  };

  render() {
    const {
      classes,
      resourceURL,
      onResourceUrlChange,
      loadedResourceURL,
      history,
      back,
      onLoadResource,
    } = this.props;
    return (
      <div className={classes.head}>
        <div className={classes.headAddressLine}>
          <TextField
            label="Collection or Manifest URL"
            type="url"
            autoComplete="on"
            value={resourceURL}
            onChange={onResourceUrlChange}
            margin="dense"
            variant="outlined"
            className={classes.urlInputStyles}
            InputProps={{
              endAdornment: history.length > 0 && (
                <InputAdornment
                  position="end"
                  onClick={this.handleHistoryOpen}
                  className={classes.historyIconHolder}
                >
                  <DefaultTooltip title="History">
                    <History />
                  </DefaultTooltip>
                </InputAdornment>
              ),
            }}
          />
          <div className={classes.loadResourceIconSpacer}>
            <DefaultTooltip title="Load URL">
              <IconButton onClick={onLoadResource}>
                <ArrowForward />
              </IconButton>
            </DefaultTooltip>
          </div>
        </div>
        <div className={classes.selectHideout}>
          <Select
            open={this.state.historyOpen}
            onClose={this.handleHistoryClose}
            onOpen={this.handleHistoryOpen}
            value={loadedResourceURL}
            onChange={back}
          >
            {history.map(url => (
              <MenuItem key={`history__${url}`} value={url}>
                {url}
              </MenuItem>
            ))}
          </Select>
        </div>
      </div>
    );
  }
}

export default withStyles(style)(NavBar);
