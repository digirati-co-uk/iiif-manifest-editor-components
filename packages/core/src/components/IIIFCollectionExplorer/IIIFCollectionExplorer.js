import * as React from 'react';
import * as PropTypes from 'prop-types';
import {
  withStyles,
  CircularProgress,
  Dialog,
  DialogTitle,
  DialogContent,
  DialogContentText,
  DialogActions,
  Button,
  Slide,
} from '@material-ui/core';

import convertToV3ifNecessary from '../../utils/IIIFUpgrader';
import NavBar from './IIIFCollectionExplorer.NavBar';
import CanvasList from './IIIFCollectionExplorer.CanvasList';
import CollectionLister from './IIIFCollectionExplorer.CollectionLister';

const isManifestOrCollection = resource =>
  resource.type === 'Collection' || resource.type === 'Manifest';

const styles = theme => ({
  root: {
    display: 'flex',
    flexDirection: 'column',
    width: '100%',
    height: '100%',
    minHeight: 300,
  },
  loadingIndicatorContainer: {
    width: '100%',
    height: '100%',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
});

class CollectionExplorer extends React.Component {
  state = {
    resourceURL: '',
    loadedResourceURL: '',
    resource: null,
    isLoading: false,
    history: [],
    error: null,
    manifestIcon: null,
    collectionIcon: null,
  };

  constructor(props) {
    super(props);
    const { url } = this.props;
    if (url && url.length) {
      this.loadResource(url);
    }
  }

  onLoadResource = ev => this.loadResource(this.state.resourceURL);
  back = ev => this.loadResource(ev.target.value);

  //TODO: CORS proxy
  loadResource = url => {
    if (url !== '' && url !== this.state.loadedResourceURL) {
      this.setState({
        isLoading: true,
      });
      fetch(url.replace(/^https?\:\/\//, '//'))
        .then(response => {
          if (!response.ok) {
            throw `[${response.status}] ${response.statusText}`;
          }
          return response;
        })
        .then(response => response.json())
        .then(resource => convertToV3ifNecessary(resource))
        .then(resource => {
          let history = this.state.history;
          const resourceUrlIndex = history.indexOf(url);
          if (resourceUrlIndex !== -1) {
            history = history.slice(0, resourceUrlIndex);
          }
          if (
            !this.props.autoSelectIfManifestFromUrl ||
            (this.props.autoSelectIfManifestFromUrl &&
              !this.props.onItemSelect(resource))
          ) {
            this.setState(
              {
                resource,
                loadedResourceURL: url,
                resourceURL: url,
                isLoading: false,
                history: history.concat([url]),
              },
              () => {
                this.props.onResourceLoaded(url);
              }
            );
          }
        })
        .catch(err =>
          this.setState({
            isLoading: false,
            error: `Failed to load ${url}.`,
          })
        );
    }
  };

  onResourceUrlChange = ev => {
    this.setState({
      resourceURL: ev.target.value,
    });
  };

  openItem = item => {
    if (item.items) {
      this.setState({
        resource: item.items,
      });
    } else if (item.id) {
      if (!this.props.onItemSelect(item)) {
        this.loadResource(item.id);
      }
    }
  };

  handleCloseErrorDialog = () => {
    this.setState({
      error: false,
    });
  };

  render() {
    const { classes, style, manifestIcon, collectionIcon } = this.props;
    const {
      history,
      loadedResourceURL,
      isLoading,
      resourceURL,
      resource,
      error,
    } = this.state;

    const collectionListExtraProps = {};
    if (manifestIcon) {
      collectionListExtraProps.manifestIcon = manifestIcon;
    }
    if (collectionIcon) {
      collectionListExtraProps.collectionIcon = collectionIcon;
    }

    const items =
      resource && isManifestOrCollection(resource) && resource.items
        ? resource.items
        : []; // TODO: add empty result set

    return (
      <div className={classes.root} style={style}>
        <NavBar
          resourceURL={resourceURL}
          loadedResourceURL={loadedResourceURL}
          history={history}
          onResourceUrlChange={this.onResourceUrlChange}
          back={this.back}
          onLoadResource={this.onLoadResource}
        />
        {isLoading ? (
          <div className={classes.loadingIndicatorContainer}>
            <CircularProgress />
          </div>
        ) : resource && resource.type === 'Manifest' ? (
          <CanvasList items={items} manifestId={resource.id} />
        ) : (
          <CollectionLister
            items={items}
            openItem={this.openItem}
            {...collectionListExtraProps}
          />
        )}
        {!!error && (
          <Dialog
            open={!!error}
            TransitionComponent={props => <Slide direction="up" {...props} />}
            onClose={this.handleCloseErrorDialog}
            aria-labelledby="error"
          >
            <DialogTitle>Error</DialogTitle>
            <DialogContent>
              <DialogContentText>{error}</DialogContentText>
            </DialogContent>
            <DialogActions>
              <Button onClick={this.handleCloseErrorDialog} color="primary">
                Ok
              </Button>
            </DialogActions>
          </Dialog>
        )}
      </div>
    );
  }
}

CollectionExplorer.propTypes = {
  url: PropTypes.string,
  onItemSelect: PropTypes.func,
  autoSelectIfManifestFromUrl: PropTypes.bool,
  onResourceLoaded: PropTypes.func,
  manifestIcon: PropTypes.any,
  collectionIcon: PropTypes.any,
};

CollectionExplorer.defaultProps = {
  url: '',
  onItemSelect: () => {},
  onResourceLoaded: () => {},
  manifestIcon: null,
  collectionIcon: null,
  autoSelectIfManifestFromUrl: false,
};

export default withStyles(styles)(CollectionExplorer);
