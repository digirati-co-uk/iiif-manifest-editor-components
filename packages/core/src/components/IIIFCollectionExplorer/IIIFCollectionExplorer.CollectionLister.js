import * as React from 'react';
import * as PropTypes from 'prop-types';

import { withStyles, Divider, Typography } from '@material-ui/core';
import { ArrowForwardIos } from '@material-ui/icons';

import IMG from './IIIFCollectionExplorer.IMG';
import LocaleString from '../LocaleString/LocaleString';
import { getCollectionThumbnail } from './IIIFCollectionExplorer.utils';

const style = theme => ({
  list: {
    display: 'flex',
    flexWrap: 'wrap',
    overflowX: 'hidden',
    overflowY: 'auto',
    position: 'relative',
  },
  listItem: {
    display: 'block',
    width: '100%',
    cursor: 'pointer',
    display: 'flex',
    flexDirection: 'row',
    flexWrap: 'wrap',
    alignItems: 'flex-start',
    justifyContent: 'flex-start',
  },
  info: {
    width: 'calc(100% - 140px)',
    padding: theme.spacing.unit,
  },
  infoLong: {
    width: 'calc(100% - 40px)',
    padding: theme.spacing.unit,
  },
  moreVert: {
    width: 40,
    minHeight: 80,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
  thumbnail: {
    width: 100,
  },
  placeholder: {
    display: 'flex',
    width: 100,
    height: 100,
    alignItems: 'center',
    justifyContent: 'center',
  },
  divider: {
    width: '100%',
  },
});

const CollectionLister = ({ classes, items, openItem }) => (
  <div className={classes.list}>
    {items.map(item => {
      const thumbnail = getCollectionThumbnail(item);
      return (
        <div
          key={item.id}
          onClick={() => openItem(item)}
          className={classes.listItem}
        >
          <div className={classes.moreVert}>
            <ArrowForwardIos />
          </div>
          <div className={thumbnail ? classes.info : classes.infoLong}>
            <Typography component="h5" variant="h5">
              <LocaleString>{item.label}</LocaleString>
            </Typography>
            <Typography variant="subtitle1" color="textSecondary">
              {item.type}
            </Typography>
          </div>
          {thumbnail && (
            <IMG src={thumbnail} alt={item.id} className={classes.thumbnail} />
          )}
          <Divider className={classes.divider} />
        </div>
      );
    })}
  </div>
);

export default withStyles(style)(CollectionLister);
