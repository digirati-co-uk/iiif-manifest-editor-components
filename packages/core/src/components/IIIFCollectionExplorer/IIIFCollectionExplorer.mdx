---
name: Collection Explorer
menu: 02. High Level Components
---

import { Playground, PropsTable } from 'docz';
import IIIFCollectionExplorer from './IIIFCollectionExplorer'; 
import { DragDropContext } from 'react-beautiful-dnd';

# Collection Explorer

<PropsTable of={IIIFCollectionExplorer} />

The following demonstation loads the `iiif universe` top level collection.

<Playground>
  <DragDropContext onDragEnd={()=>{}}>
    <div 
      style={{
        width: 500,
        height: 700,
        position: 'relative',
      }}
    >
      <IIIFCollectionExplorer url="http://ryanfb.github.io/iiif-universe/iiif-universe.json" />
    </div>
  </DragDropContext>
</Playground>


TODO:
- [ ] make it able to work without drag and drop context.
- [x] give the `virtual-list` a minimum height
- [ ] list item actions
