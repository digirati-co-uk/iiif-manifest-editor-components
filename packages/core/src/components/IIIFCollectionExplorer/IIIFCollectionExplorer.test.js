import * as React from 'react';
import IIIFCollectionExplorer from './IIIFCollectionExplorer';

describe('IIIFCollectionExplorer', () => {
  it('renders the collection explorer unchanged', () => {
    const collectionExplorer = shallow(<IIIFCollectionExplorer url={null} />);
    expect(collectionExplorer).toMatchSnapshot();
  });
});
