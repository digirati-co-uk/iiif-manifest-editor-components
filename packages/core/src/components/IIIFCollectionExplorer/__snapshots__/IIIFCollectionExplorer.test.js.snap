// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IIIFCollectionExplorer renders the collection explorer unchanged 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithStyles(CollectionExplorer)
    url={null}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "classes": Object {
        "loadingIndicatorContainer": "CollectionExplorer-loadingIndicatorContainer-2",
        "root": "CollectionExplorer-root-1",
      },
      "onItemSelect": [Function],
      "onResourceLoaded": [Function],
      "url": null,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "classes": Object {
          "loadingIndicatorContainer": "CollectionExplorer-loadingIndicatorContainer-2",
          "root": "CollectionExplorer-root-1",
        },
        "onItemSelect": [Function],
        "onResourceLoaded": [Function],
        "url": null,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
